File: test/test_sensors.erl
%%%-------------------------------------------------------------------
%%% @author borja
%%% @doc
%%%
%%% A sensor must return one of the following values:
%%% @spec actuator_name(State) ->
%%%                     {  ok, Signal, NextState} |
%%%                     {stop, Reason, NextState} 
%%% @end
%%%-------------------------------------------------------------------
-module(test_sensors).
-compile([export_all, nowarn_export_all]). %%TODO: To delete after build

%% Defined agent species
-export([]).

-define(XOR_TEST_SEQ1, [-1, +1, -1, +1]).
-define(XOR_TEST_SEQ2, [-1, -1, +1, +1]).


%%%===================================================================
%%% Defined Sensors
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc Returns a list of tuples with the record name and attributes
%% list. This is mainly used to prepare the tables in mnesia.
%% @end
%%-------------------------------------------------------------------
seq_1() -> [logical_gate].

-spec seq_1(State :: term()) -> Result when
    Result :: {  ok, Signal :: number(), NewState :: term()} |
              {stop, Reason :: string(), NewState :: term()}.
seq_1(#{seq_1:=[Signal|Sx]} = State) -> 
    In = [Signal|maps:get(in, State, [])],
    {ok, Signal, State#{in=>In,seq_1:=Sx}};
seq_1(#{seq_1:=[]} = State) -> 
    {stop, "end of training", State};
seq_1(#{} = State) -> 
    seq_1(State#{seq_1=>?XOR_TEST_SEQ1}).


%%--------------------------------------------------------------------
%% @doc Returns a list of tuples with the record name and attributes
%% list. This is mainly used to prepare the tables in mnesia.
%% @end
%%------------------------------------------------------------------
seq_2() -> [logical_gate].

-spec seq_2(State :: term()) -> Result when
    Result :: {  ok, Signal :: number(), NewState :: term()} |
              {stop, Reason :: string(), NewState :: term()}.
seq_2(#{seq_2:=[Signal|Sx]} = State) -> 
    In = [Signal|maps:get(in, State, [])],
    {ok, Signal, State#{in=>In,seq_2:=Sx}};
seq_2(#{seq_2:=[]} = State) -> 
    {stop, "end of training", State};
seq_2(#{} = State) -> 
    seq_2(State#{seq_2=>?XOR_TEST_SEQ2}).


%%====================================================================
%% Internal functions
%%====================================================================

File: lib/examples/test_sensors.ex
defmodule TestSensors do
  @moduledoc """
  An sensor must return one of the following values:
  @spec sensor_name(State) ->
                    {  ok, Signal, NextState} |
                    {stop, Reason, NextState}
  """

  ### =================================================================
  ###  Groups
  ### =================================================================
  @doc """
  Function: groups() -> [Group]
  Group = {GroupName, [SensorName :: atom()]
  """
  def groups() do
    [
      {:bool_input1, [:seq_1]},
      {:bool_input2, [:seq_2]}
    ]
  end

  ### =================================================================
  ###  Defined Sensors
  ### =================================================================

  @doc """
  Returns a boolean in an specific sequence.
  """
  def seq_1(), do: []

  @xor_seq [-1, +1, -1, +1]
  @spec seq_1(Sensor.state()) :: Sensor.result()
  def seq_1(%{seq_1: [signal | sx], i1: _} = state) do
    {:ok, signal, %{state | i1: signal, seq_1: sx.()}}
  end

  def seq_1(%{i1: _} = state), do: seq_1(Map.put(state, :seq_1, lazy_seq(@xor_seq)))
  def seq_1(%{} = state), do: seq_1(Map.put(state, :i1, []))

  @doc """
  Returns a boolean in an specific sequence.
  """
  def seq_2(), do: []

  @xor_seq [-1, -1, +1, +1]
  @spec seq_2(Sensor.state()) :: Sensor.result()
  def seq_2(%{seq_2: [signal | sx], i2: _} = state) do
    {:ok, signal, %{state | i2: signal, seq_2: sx.()}}
  end

  def seq_2(%{i2: _} = state), do: seq_2(Map.put(state, :seq_2, lazy_seq(@xor_seq)))
  def seq_2(%{} = state), do: seq_2(Map.put(state, :i2, []))

  ### =================================================================
  ###  Internal functions
  ### =================================================================

  def lazy_seq(cycle) do
    lazy_queue(:queue.from_list(cycle))
  end

  defp lazy_queue(queue) do
    {{:value, h}, queue} = :queue.out(queue)
    [h | fn -> lazy_queue(:queue.in(h, queue)) end]
  end
end
